### Задание 1

Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.

Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки 
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [9]:
import pandas as pd
import numpy as np
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_boston

In [10]:
boston = load_boston()
data = boston.data
target = boston.target
feature_names = boston.feature_names

In [11]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [12]:
X = pd.DataFrame(data, columns=feature_names)
Y = pd.DataFrame(target, columns=['price'])

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.30, random_state=42)

In [20]:
from sklearn.linear_model import LinearRegression

In [21]:
lr = LinearRegression()
lr.fit(X_train, Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [22]:
Y_pred = lr.predict(X_test)
check_test = pd.DataFrame({'Y_test': Y_test['price'], 'Y_pred': Y_pred.flatten()}, columns=['Y_test', 'Y_pred'])

In [26]:
check_test.head(10)

Y_test     Y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

In [23]:
from sklearn.metrics import r2_score

In [24]:
r2_score(Y_test, Y_pred)

0.711226005748496

### Задание 2

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000, 
max_depth должен быть равен 12 и random_state сделайте равным 42.

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, 
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [25]:
from sklearn.ensemble import RandomForestRegressor

In [26]:
model = RandomForestRegressor(max_depth=12, random_state=42, n_estimators=1000)

In [27]:
model.fit(X_train, Y_train.values[:, 0])

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=12,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=1000, n_jobs=None,
           oob_score=False, random_state=42, verbose=0, warm_start=False)

In [28]:
Y_pred = model.predict(X_test)
check_test = pd.DataFrame({'Y_test': Y_test['price'], 'Y_pred': Y_pred.flatten()}, columns=['Y_test', 'Y_pred'])
check_test.head(10)

Y_test     Y_pred
173    23.6  22.846138
274    32.4  31.156114
491    13.6  16.297226
72     22.8  23.821036
452    16.1  17.212148
76     20.0  21.820092
316    17.8  19.866369
140    14.0  14.759938
471    19.6  21.235224
500    16.8  20.883103

In [29]:
r2_score(Y_test, Y_pred)

0.8749965273218174

<b>Какая модель в данном случае работает лучше? </b> <br>
С использованием линейной регрессии <b>r2 = 0.711,</b> <br>
С использованием RandomForestRegressor <b>r2 = 0.875</b><br>
В данном случае RandomForestRegressor работает лучше. 

### Задание 3

Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.

С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.

In [30]:
from sklearn.ensemble import RandomForestRegressor

In [31]:
?RandomForestRegressor

feature_importances_ : array of shape = [n_features]
    The feature importances (the higher, the more important the feature).

In [215]:
importances = model.feature_importances_
print(importances)
importances.sum()

[0.03211748 0.00154999 0.0070941  0.0011488  0.01436832 0.40270459
 0.01424477 0.06403265 0.00496762 0.01169177 0.01808961 0.0123114
 0.41567892]


0.9999999999999994

In [214]:
indices = np.argsort(importances)[::-1]
indices

array([12,  5,  7,  0, 10,  4,  6, 11,  9,  2,  8,  1,  3])

Два признака показывают наибольшую важность: 12 и 5.

In [220]:
X_train.iloc[:, [12, 5]].head(10)

LSTAT     RM
5     5.21  6.430
116  12.04  6.176
45   10.21  5.682
16    6.58  5.935
468  18.13  5.926
360   7.79  6.398
3     2.94  6.998
405  22.98  5.683
185  13.15  6.153
60   13.15  5.741

### Задание 4

В этом задании мы будем работать с датасетом, с которым мы уже 
знакомы по домашнему заданию по библиотеке Matplotlib, это датасет Credit Card Fraud Detection.

Для этого датасета мы будем решать задачу классификации - будем определять, 
какие из транзакциции по кредитной карте являются мошенническими.

Данный датасет сильно несбалансирован (так как случаи мошенничества относительно редки),
так что применение метрики accuracy не принесет пользы и не поможет выбрать лучшую модель.
Мы будем вычислять AUC, то есть площадь под кривой ROC.

Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.

Загрузите датасет creditcard.csv и создайте датафрейм df.
С помощью метода value_counts с аргументом normalize=True убедитесь в том, что выборка несбалансирована.
Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет ли в них пропусков.

Примените следующую настройку, чтобы можно было просматривать все столбцы датафрейма:
pd.options.display.max_columns = 100.

Просмотрите первые 10 строк датафрейма df.

Создайте датафрейм X из датафрейма df, исключив столбец Class.
Создайте объект Series под названием y из столбца Class.

Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split, используя аргументы: test_size=0.3, random_state=100, stratify=y.
У вас должны получиться объекты X_train, X_test, y_train и y_test.
Просмотрите информацию о их форме.

Для поиска по сетке параметров задайте такие параметры:
parameters = [{'n_estimators': [10, 15], 
'max_features': np.arange(3, 5),
'max_depth': np.arange(4, 7)}]

Создайте модель GridSearchCV со следующими аргументами:
estimator=RandomForestClassifier(random_state=100), 
param_grid=parameters,
scoring='roc_auc',
cv=3.

Обучите модель на тренировочном наборе данных (может занять несколько минут).

Просмотрите параметры лучшей модели с помощью атрибута best_params_.

Предскажите вероятности классов с помощью полученнной модели и метода predict_proba.
Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и запишите в массив y_pred_proba.

Из модуля sklearn.metrics импортируйте метрику roc_auc_score.
Вычислите AUC на тестовых данных и сравните с результатом, 
полученным на тренировочных данных, используя в качестве аргументов
массивы y_test и y_pred_proba.

In [32]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
import pandas as pd
import matplotlib.style as style
df = pd.read_csv('creditcard.csv')
pd.DataFrame(df['Class'].value_counts(normalize=True))

Class
0  0.998273
1  0.001727

Выборка несбалансированная, потому что объектов одного класса значительно меньше, чем объектов другого (количество объектов класса мошеннических операций составляет 0,173% от количества объектов немошеннических операций).

In [228]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
Time      284807 non-null float64
V1        284807 non-null float64
V2        284807 non-null float64
V3        284807 non-null float64
V4        284807 non-null float64
V5        284807 non-null float64
V6        284807 non-null float64
V7        284807 non-null float64
V8        284807 non-null float64
V9        284807 non-null float64
V10       284807 non-null float64
V11       284807 non-null float64
V12       284807 non-null float64
V13       284807 non-null float64
V14       284807 non-null float64
V15       284807 non-null float64
V16       284807 non-null float64
V17       284807 non-null float64
V18       284807 non-null float64
V19       284807 non-null float64
V20       284807 non-null float64
V21       284807 non-null float64
V22       284807 non-null float64
V23       284807 non-null float64
V24       284807 non-null float64
V25       284807 non-null float64
V26  

Пропусков нет.

In [229]:
pd.options.display.max_columns = 100

In [230]:
df.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9       V10       V11       V12       V13       V14  \
0  0.098698  0.363787  0.090794 -0.551600 -0.617801 -0.991390 -0.311169   
1  0.085102 -0.255425 -0.166974  1.612727  1.065235  0.489095 -0.143772   
2  0.247676 -1.514654  0.207643  0.624501  0.066084  0.717293 -0.165946   
3  0.377436 -1.387024 -0.054952 -0.226487  0.178228  0.507757 -0.287924   
4 -0.270533  0.817739  0.753074 -0.822843  0.538196  1.345852 -1.119670   
5  0.260314 -0.568671 -0.371407  1.341262  0.359894 -0.358091 -0.137134   
6  0.081213  0.464960 -0.099254 -1.416907 -0.153826 -0.751063  0.167372   
7 -3.807864  0.615375  1.249376 -0.619468  0.291474  1.757964 -1.323865   
8  0.851084 -0.392048 -0.410430 -0.705117 -0.110452 -0.286254  0.074355   
9  0.069539 -0.736727 -0.366846  1.017614  0.836390  1.006844 -0.443523   

        V15       V16       V17       V18       V19       V20       V21  \
0  1.468177 -0.470401  0.207971  0.025791  0.403993  0.251412 -0.018307   
1  0.635558  0.463917 -0.114805 -0.183361 -0.145783 -0.069083 -0.225775   
2  2.345865 -2.890083  1.109969 -0.121359 -2.261857  0.524980  0.247998   
3 -0.631418 -1.059647 -0.684093  1.965775 -1.232622 -0.208038 -0.108300   
4  0.175121 -0.451449 -0.237033 -0.038195  0.803487  0.408542 -0.009431   
5  0.517617  0.401726 -0.058133  0.068653 -0.033194  0.084968 -0.208254   
6  0.050144 -0.443587  0.002821 -0.611987 -0.045575 -0.219633 -0.167716   
7  0.686133 -0.076127 -1.222127 -0.358222  0.324505 -0.156742  1.943465   
8 -0.328783 -0.210077 -0.499768  0.118765  0.570328  0.052736 -0.073425   
9  0.150219  0.739453 -0.540980  0.476677  0.451773  0.203711 -0.246914   

        V22       V23       V24       V25       V26       V27       V28  \
0  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
5 -0.559825 -0.026398 -0.371427 -0.232794  0.105915  0.253844  0.081080   
6 -0.270710 -0.154104 -0.780055  0.750137 -0.257237  0.034507  0.005168   
7 -1.015455  0.057504 -0.649709 -0.415267 -0.051634 -1.206921 -1.085339   
8 -0.268092 -0.204233  1.011592  0.373205 -0.384157  0.011747  0.142404   
9 -0.633753 -0.120794 -0.385050 -0.069733  0.094199  0.246219  0.083076   

   Amount  Class  
0  149.62      0  
1    2.69      0  
2  378.66      0  
3  123.50      0  
4   69.99      0  
5    3.67      0  
6    4.99      0  
7   40.80      0  
8   93.20      0  
9    3.68      0

In [262]:
y = df.Class
X = pd.DataFrame(df.iloc[:, 0:-1])

In [261]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100, stratify=y)

In [263]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(199364, 30) (85443, 30) (199364,) (85443,)


In [265]:
parameters = [{'n_estimators': [10, 15], 
'max_features': np.arange(3, 5),
'max_depth': np.arange(4, 7)}]

clf = GridSearchCV(estimator=RandomForestClassifier(random_state=100), 
param_grid=parameters,
scoring='roc_auc',
cv=3)

clf.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise-deprecating',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators='warn', n_jobs=None,
            oob_score=False, random_state=100, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid=[{'n_estimators': [10, 15], 'max_features': array([3, 4]), 'max_depth': array([4, 5, 6])}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [266]:
clf.best_params_

{'max_depth': 6, 'max_features': 3, 'n_estimators': 15}

In [282]:
y_pred = clf.predict_proba(X_test)
y_pred_proba = y_pred[:, 1]
y_pred_proba

array([0.00092917, 0.00029521, 0.00028215, ..., 0.00028215, 0.0006822 ,
       0.01246098])

In [280]:
from sklearn.metrics import roc_auc_score

In [283]:
roc_auc_score(y_test, y_pred_proba)

0.9462664156037156

## Дополнительные задания:

1). Загрузите датасет Wine из встроенных датасетов sklearn.datasets с помощью функции load_wine в переменную data.

In [33]:
from sklearn.datasets import load_wine
wine = load_wine()

2). Полученный датасет не является датафреймом. Это структура данных, имеющая ключи аналогично словарю. Просмотрите тип данных этой структуры данных и создайте список data_keys, содержащий ее ключи.

In [34]:
data_keys = wine.keys()
data_keys

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

3). Просмотрите данные, описание и названия признаков в датасете. Описание нужно вывести в виде привычного, аккуратно оформленного текста, без обозначений переноса строки, но с самими переносами и т.д.

In [35]:
for line in wine.DESCR.split('/n'):
    print(line)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

4). Сколько классов содержит целевая переменная датасета? Выведите названия классов.

In [36]:
target = wine.target
print(len(wine['target_names']))
wine['target_names']

3


array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [37]:
feature_names = wine.feature_names
feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

5). На основе данных датасета (они содержатся в двумерном массиве Numpy) и названий признаков создайте датафрейм под названием X.

In [38]:
X = pd.DataFrame(data, columns=feature_names)
X.head(10)

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0  0.00632        18.0  2.31                0.0      0.538          6.575   
1  0.02731         0.0  7.07                0.0      0.469          6.421   
2  0.02729         0.0  7.07                0.0      0.469          7.185   
3  0.03237         0.0  2.18                0.0      0.458          6.998   
4  0.06905         0.0  2.18                0.0      0.458          7.147   
5  0.02985         0.0  2.18                0.0      0.458          6.430   
6  0.08829        12.5  7.87                0.0      0.524          6.012   
7  0.14455        12.5  7.87                0.0      0.524          6.172   
8  0.21124        12.5  7.87                0.0      0.524          5.631   
9  0.17004        12.5  7.87                0.0      0.524          6.004   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        65.2                4.0900              1.0            296.0  15.3   
1        78.9                4.9671              2.0            242.0  17.8   
2        61.1                4.9671              2.0            242.0  17.8   
3        45.8                6.0622              3.0            222.0  18.7   
4        54.2                6.0622              3.0            222.0  18.7   
5        58.7                6.0622              3.0            222.0  18.7   
6        66.6                5.5605              5.0            311.0  15.2   
7        96.1                5.9505              5.0            311.0  15.2   
8       100.0                6.0821              5.0            311.0  15.2   
9        85.9                6.5921              5.0            311.0  15.2   

   od280/od315_of_diluted_wines  proline  
0                        396.90     4.98  
1                        396.90     9.14  
2                        392.83     4.03  
3                        394.63     2.94  
4                        396.90     5.33  
5                        394.12     5.21  
6                        395.60    12.43  
7                        396.90    19.15  
8                        386.63    29.93  
9                        386.71    17.10

6). Выясните размер датафрейма X и установите, имеются ли в нем пропущенные значения.

In [336]:
X.shape

(178, 13)

In [339]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
alcohol                         178 non-null float64
malic_acid                      178 non-null float64
ash                             178 non-null float64
alcalinity_of_ash               178 non-null float64
magnesium                       178 non-null float64
total_phenols                   178 non-null float64
flavanoids                      178 non-null float64
nonflavanoid_phenols            178 non-null float64
proanthocyanins                 178 non-null float64
color_intensity                 178 non-null float64
hue                             178 non-null float64
od280/od315_of_diluted_wines    178 non-null float64
proline                         178 non-null float64
dtypes: float64(13)
memory usage: 18.2 KB


<b>Вывод:</b> <br>во всех колонках по 178 значений, значит пропущенных нет. 

7). Добавьте в датафрейм поле с классами вин в виде чисел, имеющих тип данных numpy.int64. Название поля - 'target'.

In [40]:
import numpy
X['target'] = pd.Series(target, dtype=numpy.int64)
X.head(5)

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0  0.00632        18.0  2.31                0.0      0.538          6.575   
1  0.02731         0.0  7.07                0.0      0.469          6.421   
2  0.02729         0.0  7.07                0.0      0.469          7.185   
3  0.03237         0.0  2.18                0.0      0.458          6.998   
4  0.06905         0.0  2.18                0.0      0.458          7.147   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        65.2                4.0900              1.0            296.0  15.3   
1        78.9                4.9671              2.0            242.0  17.8   
2        61.1                4.9671              2.0            242.0  17.8   
3        45.8                6.0622              3.0            222.0  18.7   
4        54.2                6.0622              3.0            222.0  18.7   

   od280/od315_of_diluted_wines  proline  target  
0                        396.90     4.98     0.0  
1                        396.90     9.14     0.0  
2                        392.83     4.03     0.0  
3                        394.63     2.94     0.0  
4                        396.90     5.33     0.0

8). Постройте матрицу корреляций для всех полей X. Дайте полученному датафрейму название X_corr.

In [52]:
X_corr = X.corr()
X_corr

alcohol  malic_acid       ash  \
alcohol                       1.000000   -0.200469  0.406583   
malic_acid                   -0.200469    1.000000 -0.533828   
ash                           0.406583   -0.533828  1.000000   
alcalinity_of_ash            -0.055892   -0.042697  0.062938   
magnesium                     0.420972   -0.516604  0.763651   
total_phenols                -0.219247    0.311991 -0.391676   
flavanoids                    0.352734   -0.569537  0.644779   
nonflavanoid_phenols         -0.379670    0.664408 -0.708027   
proanthocyanins               0.625505   -0.311948  0.595129   
color_intensity               0.582764   -0.314563  0.720760   
hue                           0.289946   -0.391679  0.383248   
od280/od315_of_diluted_wines -0.385064    0.175520 -0.356977   
proline                       0.455621   -0.412995  0.603800   
target                        0.469069   -0.254846  0.635748   

                              alcalinity_of_ash  magnesium  total_phenols  \
alcohol                               -0.055892   0.420972      -0.219247   
malic_acid                            -0.042697  -0.516604       0.311991   
ash                                    0.062938   0.763651      -0.391676   
alcalinity_of_ash                      1.000000   0.091203       0.091251   
magnesium                              0.091203   1.000000      -0.302188   
total_phenols                          0.091251  -0.302188       1.000000   
flavanoids                             0.086518   0.731470      -0.240265   
nonflavanoid_phenols                  -0.099176  -0.769230       0.205246   
proanthocyanins                       -0.007368   0.611441      -0.209847   
color_intensity                       -0.035587   0.668023      -0.292048   
hue                                   -0.121515   0.188933      -0.355501   
od280/od315_of_diluted_wines           0.048788  -0.380051       0.128069   
proline                               -0.053929   0.590879      -0.613808   
target                                 0.277968   0.589181      -0.019265   

                              flavanoids  nonflavanoid_phenols  \
alcohol                         0.352734             -0.379670   
malic_acid                     -0.569537              0.664408   
ash                             0.644779             -0.708027   
alcalinity_of_ash               0.086518             -0.099176   
magnesium                       0.731470             -0.769230   
total_phenols                  -0.240265              0.205246   
flavanoids                      1.000000             -0.747881   
nonflavanoid_phenols           -0.747881              1.000000   
proanthocyanins                 0.456022             -0.494588   
color_intensity                 0.506456             -0.534432   
hue                             0.261515             -0.232471   
od280/od315_of_diluted_wines   -0.273534              0.291512   
proline                         0.602339             -0.496996   
target                          0.402879             -0.672979   

                              proanthocyanins  color_intensity       hue  \
alcohol                              0.625505         0.582764  0.289946   
malic_acid                          -0.311948        -0.314563 -0.391679   
ash                                  0.595129         0.720760  0.383248   
alcalinity_of_ash                   -0.007368        -0.035587 -0.121515   
magnesium                            0.611441         0.668023  0.188933   
total_phenols                       -0.209847        -0.292048 -0.355501   
flavanoids                           0.456022         0.506456  0.261515   
nonflavanoid_phenols                -0.494588        -0.534432 -0.232471   
proanthocyanins                      1.000000         0.910228  0.464741   
color_intensity                      0.910228         1.000000  0.460853   
hue                                  0.464741         0.460853  1.000000   
od280/od315

9). Создайте список high_corr из признаков, корреляция которых с полем target по абсолютному значению превышает 0.5 (причем, само поле target не должно входить в этот список).

In [68]:
high_corr = []
for i in range(len(X_corr['target']) - 1):
    if X_corr['target'][i] >=0.5 or X_corr['target'][i]<= -0.5:
       high_corr.append(X_corr.columns[i]) 
high_corr

['ash', 'magnesium', 'nonflavanoid_phenols', 'color_intensity']

10). Удалите из датафрейма X поле с целевой переменной. Для всех признаков, названия которых содержатся в списке high_corr, вычислите квадрат их значений и добавьте в датафрейм X соответствующие поля с суффиксом '_2', добавленного к первоначальному названию признака. Итоговый датафрейм должен содержать все поля, которые, были в нем изначально, а также поля с признаками из списка high_corr, возведенными в квадрат. Выведите описание полей датафрейма X с помощью метода describe.

In [ ]:
del X['target']

In [73]:
X.head(3)

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0  0.00632        18.0  2.31                0.0      0.538          6.575   
1  0.02731         0.0  7.07                0.0      0.469          6.421   
2  0.02729         0.0  7.07                0.0      0.469          7.185   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        65.2                4.0900              1.0            296.0  15.3   
1        78.9                4.9671              2.0            242.0  17.8   
2        61.1                4.9671              2.0            242.0  17.8   

   od280/od315_of_diluted_wines  proline  
0                        396.90     4.98  
1                        396.90     9.14  
2                        392.83     4.03

In [79]:
X.shape[1]

13

In [81]:
X.columns[1]

'malic_acid'

In [98]:
for i in range(X.shape[1]):
    if X.columns[i] in high_corr:
        X[f'{X.columns[i]}_2'] = pd.Series((X[X.columns[i]]**2))

In [94]:
X.head(2)

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0  0.00632        18.0  2.31                0.0      0.538          6.575   
1  0.02731         0.0  7.07                0.0      0.469          6.421   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        65.2                4.0900              1.0            296.0  15.3   
1        78.9                4.9671              2.0            242.0  17.8   

   od280/od315_of_diluted_wines  proline    ash_2  magnesium_2  \
0                         396.9     4.98   5.3361     0.289444   
1                         396.9     9.14  49.9849     0.219961   

   nonflavanoid_phenols_2  color_intensity_2  
0               16.728100            87616.0  
1               24.672082            58564.0

In [97]:
X.describe()

alcohol  malic_acid         ash  alcalinity_of_ash   magnesium  \
count  506.000000  506.000000  506.000000         506.000000  506.000000   
mean     3.613524   11.363636   11.136779           0.069170    0.554695   
std      8.601545   23.322453    6.860353           0.253994    0.115878   
min      0.006320    0.000000    0.460000           0.000000    0.385000   
25%      0.082045    0.000000    5.190000           0.000000    0.449000   
50%      0.256510    0.000000    9.690000           0.000000    0.538000   
75%      3.677083   12.500000   18.100000           0.000000    0.624000   
max     88.976200  100.000000   27.740000           1.000000    0.871000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count     506.000000  506.000000            506.000000       506.000000   
mean        6.284634   68.574901              3.795043         9.549407   
std         0.702617   28.148861              2.105710         8.707259   
min         3.561000    2.900000              1.129600         1.000000   
25%         5.885500   45.025000              2.100175         4.000000   
50%         6.208500   77.500000              3.207450         5.000000   
75%         6.623500   94.075000              5.188425        24.000000   
max         8.780000  100.000000             12.126500        24.000000   

       color_intensity         hue  od280/od315_of_diluted_wines     proline  \
count       506.000000  506.000000                    506.000000  506.000000   
mean        408.237154   18.455534                    356.674032   12.653063   
std         168.537116    2.164946                     91.294864    7.141062   
min         187.000000   12.600000                      0.320000    1.730000   
25%         279.000000   17.400000                    375.377500    6.950000   
50%         330.000000   19.050000                    391.440000   11.360000   
75%         666.000000   20.200000                    396.225000   16.955000   
max         711.000000   22.000000                    396.900000   37.970000   

            ash_2  magnesium_2  nonflavanoid_phenols_2  color_intensity_2  
count  506.000000   506.000000              506.000000         506.000000  
mean   170.999269     0.321088               18.827601      195006.197628  
std    171.087497     0.139213               21.136892      157613.919181  
min      0.211600     0.148225                1.275996       34969.000000  
25%     26.936100     0.201601                4.410735       77841.000000  
50%     93.896100     0.289444               10.287804      108900.000000  
75%    327.610000     0.389376               26.921407      443556.000000  
max    769.507600     0.758641              147.052002      505521.000000